doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -0,0 +1,969 @@
+   corresponding generated expression.
+
+   For instance, assume we perform:
+
+   :n:`Ltac2 Notation "foo" c(thunk(constr)) ids(list0(ident)) := Bar.f c ids.`
+
+   Then the following expression
+
+   :n:`let y := @X in foo (nat -> nat) x $y`
+
+   will expand at parsing time to
+
+   :n:`let y := @X in`
+   :n:`let c := fun () => constr:(nat -> nat) with ids := [@x; y] in Bar.f c ids`
+
+   Beware that the order of evaluation of multiple let-bindings is not specified,
+   so that you may have to resort to thunking to ensure that side-effects are
+   performed at the right time.
+
+Abbreviations
+~~~~~~~~~~~~~
+
+.. cmdv:: Ltac2 Notation @ident := @term
+
+  This command introduces a special kind of notations, called abbreviations,
+  that is designed so that it does not add any parsing rules. It is similar in
+  spirit to Coq abbreviations, insofar as its main purpose is to give an
+  absolute name to a piece of pure syntax, which can be transparently referred
+  by this name as if it were a proper definition.
+
+  The abbreviation can then be manipulated just as a normal Ltac2 definition,
+  except that it is expanded at internalization time into the given expression.
+  Furthermore, in order to make this kind of construction useful in practice in
+  an effectful language such as Ltac2, any syntactic argument to an abbreviation
+  is thunked on-the-fly during its expansion.
+
+For instance, suppose that we define the following.
+
+:n:`Ltac2 Notation foo := fun x => x ().`
+
+Then we have the following expansion at internalization time.
+
+:n:`foo 0 ↦ (fun x => x ()) (fun _ => 0)`
+
+Note that abbreviations are not typechecked at all, and may result in typing
+errors after expansion.
+
+Evaluation
+----------
+
+Ltac2 features a toplevel loop that can be used to evaluate expressions.
+
+.. cmd:: Ltac2 Eval @term
+   :name: Ltac2 Eval
+
+   This command evaluates the term in the current proof if there is one, or in the
+   global environment otherwise, and displays the resulting value to the user
+   together with its type. This function is pure in the sense that it does not
+   modify the state of the proof, and in particular all side-effects are discarded.
+
+Debug
+-----
+
+.. opt:: Ltac2 Backtrace
+
+   When this option is set, toplevel failures will be printed with a backtrace.
+
+Compatibility layer with Ltac1
+------------------------------
+
+Ltac1 from Ltac2
+~~~~~~~~~~~~~~~~
+
+Simple API
+++++++++++
+
+One can call Ltac1 code from Ltac2 by using the :n:`ltac1` quotation. It parses
+a Ltac1 expression, and semantics of this quotation is the evaluation of the
+corresponding code for its side effects. In particular, in cannot return values,
+and the quotation has type :n:`unit`.
+
+Beware, Ltac1 **cannot** access variables from the Ltac2 scope. One is limited
+to the use of standalone function calls.
+
+Low-level API
++++++++++++++
+
+There exists a lower-level FFI into Ltac1 that is not recommended for daily use,
```

Comment:
```suggestion
There exists a lower-level API into Ltac1 that is not recommended for daily use,
```

Reply:
Why not FFI? That's the designated nomenclature for a trans-language API.

Reply:
I thought it was a typo.  Even so, it's probably not a term that most readers would be familiar with.  This is the only use in the document, it's use is not essential here (compared to "API") and there's no definition.

Documentation should be written to be simple to understand, not introducing additional terms unless it's really necessary.  Terms that may not be understood by, say, half the readers should be avoided or defined.  I think that's different from academic writing where use of the latest jargon is probably expected.

Reply:
FFI is by no means academic IMHO, even JavaScript programmers know the term :) It sounds as standard as API to me.

Reply:
I didn't say it's an academic term.  I never ran across it when I've written JavaScript.  How many of our users know JavaScript and would know the term?

Reply:
It's not specific to JavaScript, @maximedenes mentioned this as an example. *FFI* is a term whose usage frequency is on par with *API*, for the good reason that they are notions very similar in nature.

Reply:
Our opinions are subjective.  I've worked in Silicon Valley a long while and never encountered the term.  Terminology choices vary across groups and other categories.  Let's not get hung up on this now.

Of course, "JavaScript programmers" is not equivalent to "inexperienced/junior programmers".  In some ways JavaScript is harder to use than OCaml (e.g. JS has minimal type checking).  Delivering significant applications that work across multiple browsers can be quite complex.

Reply:
"FFI" is common, I believe, when developers use two languages that have to interface, and more common if those two languages are at different places in the high-level / low-level spectrum.  So you might talk about a C FFI for OCaml/Haskell/Python/Perl/etc, or a Python FFI for Perl, or a Lua FFI for TeX.

According to [Wikipedia](https://en.wikipedia.org/wiki/Foreign_function_interface):
> The term comes from the specification for Common Lisp, which explicitly refers to the language features for inter-language calls as such;[1] the term is also used officially by the Haskell[2] and Python programming languages.[3] Other languages use other terminology: the Ada programming language talks about "language bindings", while Java refers to its FFI as the JNI (Java Native Interface) or JNA (Java Native Access). Foreign function interface has become generic terminology for mechanisms which provide such services.

Reply:
> Of course, "JavaScript programmers" is not equivalent to "inexperienced/junior programmers".

FTR, I never meant that.